### PR TITLE
Fix 'Cannot find column' in read-in-order optimization with ARRAY JOIN

### LINKED ISSUE #48682: Cannot find column less(id, 100) in ActionsDAG result. (UNKNOWN_IDENTIFIER)
Good afternoon, I ran into such a problem that I'm trying to select from a view and I'm catching an error. I found how to fix the error - turn off the optimization "setting optimize_read_in_order" , but in my opinion this setting should not throw an exception.

version Clickhouse 23.3.1.2823-lts

example:
https://fiddle.clickhouse.com/66ae12fb-11ee-4684-b07f-531119686419

or
```
drop table if exists default.test_array_joins
;
create table default.test_array_joins
(
    id UInt64 default rowNumberInAllBlocks() + 1,
    arr_1 Array(String),
    arr_2 Array(String),
    arr_3 Array(String),
    arr_4 Array(String)
) engine = MergeTree order by id
;
insert into default.test_array_joins (id,arr_1, arr_2, arr_3, arr_4)
SELECT number,array(randomPrintableASCII(3)),array(randomPrintableASCII(3)),array(randomPrintableASCII(3)),array(randomPrintableASCII(3))
from numbers(1000)
;
create  or replace view default.v4test_array_joins as SELECT * from default.test_array_joins where id != 10
;
select * from default.v4test_array_joins
array join columns('^arr')
where match(arr_4,'a')
  and id < 100
ORDER by id limit 1 by arr_1 settings optimize_read_in_order = 1
```

Another interesting case is if in the where clause add a known true condition first. It's work without errors!)
```
select * from default.v4test_array_joins
array join columns('^arr')
where true and match(arr_4,'a')
  and id < 100
ORDER by id limit 1 by arr_1 settings optimize_read_in_order = 1
```

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix `Cannot find column` in `query_plan_read_in_order` optimization with `ARRAY JOIN`. Fixes #48682

### CODE REVIEW

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp [hunk 1]
```diff
@@ -175,6 +175,8 @@ static void appendExpression(ActionsDAGPtr & dag, const ActionsDAGPtr & expressi
         dag->mergeInplace(std::move(*expression->clone()));
     else
         dag = expression->clone();
+
+    dag->projectInput(false);
```
Comment (KochetovNicolai) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:179]: The fix is here.

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit 5025edb4723cef7f2387ce6534ea7ab9865ebaf0 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51746/5025edb4723cef7f2387ce6534ea7ab9865ebaf0/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by qoega** [comment]
@KochetovNicolai failure is related

**Review by nickitat — APPROVED** [review]

**Comment by KochetovNicolai** [comment]
Integration tests (asan) : `services.schema-registry-auth.ports contains an invalid type,`

**Comment by KochetovNicolai** [comment]
(asan, analyzer) [2/6]  - #52375

**Comment by KochetovNicolai** [comment]
[test_s3_engine_heavy_write_check_mem](https://play.clickhouse.com/play?user=play#U0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIGNoZWNrX3N0YXR1cywgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnT0snCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwogICAgQU5EIHRlc3RfbmFtZSBsaWtlICcldGVzdF9zM19lbmdpbmVfaGVhdnlfd3JpdGVfY2hlY2tfbWVtJScKT1JERVIgQlkgY2hlY2tfc3RhcnRfdGltZSBkZXNjLCBjaGVja19uYW1lLCB0ZXN0X25hbWU=) - flaky

https://github.com/ClickHouse/ClickHouse/issues/55270

[02263_lazy_mark_load](https://play.clickhouse.com/play?user=play#U0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIGNoZWNrX3N0YXR1cywgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnT0snCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwogICAgQU5EIHRlc3RfbmFtZSBsaWtlICclMDIyNjNfbGF6eV9tYXJrX2xvYWQlJwpPUkRFUiBCWSBjaGVja19zdGFydF90aW1lIGRlc2MsIGNoZWNrX25hbWUsIHRlc3RfbmFtZQ==) - flaky

[02015_executable_user_defined_functions](U0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIGNoZWNrX3N0YXR1cywgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnT0snCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwogICAgQU5EIHRlc3RfbmFtZSBsaWtlICclMDIwMTVfZXhlY3V0YWJsZV91c2VyX2RlZmluZWRfZnVuY3Rpb25zJScKT1JERVIgQlkgY2hlY2tfc3RhcnRfdGltZSBkZXNjLCBjaGVja19uYW1lLCB0ZXN0X25hbWU) - flaky

**Review by nickitat — APPROVED** [review]

**Comment by nickitat** [comment]
it isn't relevant, right?

https://s3.amazonaws.com/clickhouse-test-reports/51746/105d16474e52022a29b5be22931e3f55e3d8029a/performance_comparison_[1_4]/report.html

**Comment by KochetovNicolai** [comment]
Yes, this is still relevant. I need to fix tests.

**Comment by KochetovNicolai** [comment]
Everything is green. Merging.

**Pull request merged**